core/src/polling_monitor/mod.rs

```diff
@@ -124,7 +171,16 @@ where
                                     "error" => format!("{:#}", e),
                                     "object_id" => id.to_string());
                         metrics.errors.inc();
-                        queue.push_back(id);
+
+                        // Requests that return errors could mean there is a permanent issue with
+                        // fetching the given item, or could signal the endpoint is overloaded.
+                        // Either way a backoff makes sense.
+                        let queue = queue.cheap_clone();
+                        let backoff = backoffs.next_backoff(id.clone());
+                        graph::spawn(async move {
```

Comment:
this creates essentially an unbounded number of tasks. How does this affect the unconstrained task the runner is running on? I kinda would prefer a pending queue we could fetch stuff on a background task. This way it's easier to track and it's memory bound. I may be wrong here but the task seems too expensive for what is happening here.

Reply:
If one of these waits forever then the `queue` would not be dropped even if the polling task ends, but that's not the case since there is a limit to the sleep time.

The tokio scheduler is [very efficient](https://tokio.rs/blog/2019-10-scheduler). And there is no fundamental reason why using tasks would be less efficient than rolling our own way of waiting on concurrent futures. There is `FuturesUnordered`, which in essence a single-task executor, but that's more suitable when there is a list of futures to be executed.

If we wanted to use a single task it might look like:
```rust
let (tx, rx) = mpsc::channel(1);
// Pull from the channel and simultaneously wait on up to a bounded number of futures.
// Uses `FuturesUnordered` under the hood.
task::spawn(rx.buffered_unordered(1_000_000).for_each(|_| ()));

// do `tx.push` this instead of `task::spawn`.
// Blocks if we've hit the `buffered_unordered` limit.
tx.push(async {
 backoff.await;
 queue.push_back(id);
}).await
```

This gives us the possibility of bounding the max number of timeouts being waited on, but I'm not sure that's particularly helpful in our case.

Reply:
What ultimately protects this from growing without bound is the max data sources setting.

Reply:
it's up to you, personally I don't think creating an unbounded number of tasks is great, especially because we can have 100s of SGs on the same runtime with each of them having an unpredictable amount of files. Putting that type of loads on the runtime, even if it's efficient, seems unnecessary but I don't know tokio internals very well so I'll leave it with you to decide.

That example above I guess would work but IIUIC if you have a high backoff time, all the following will wait that time so the last to be processed would be the sum of all the backoffs? I was thinking of place an item that looks like 
```
{
  id
  not_before: i64
}
```

Which would allow quickly scanning through the buffer and just queueing all the ready ones.

Reply:
The `unordered` aspect would avoid one backoff blocking the other. I don't really want to complicate the queue by managing the backoff directly on it, that would probably result in removing items from the middle of the queue which isn't great for performance.